controllers/paddlejob_controller.go

```diff
@@ -294,6 +311,49 @@ func (r *PaddleJobReconciler) createResource(ctx context.Context, pdj *pdv1.Padd
 
 }
 
+func (r *PaddleJobReconciler) allocHostPortForJob(ctx context.Context, pdj *pdv1.PaddleJob) bool {
+	if pdj.Spec.Intranet != pdv1.HostNetwork {
+		return false
+	}
+	if port, ok := pdj.ObjectMeta.Annotations[hostPort]; ok {
+		iport, err := strconv.Atoi(port)
+		if err == nil && iport > r.HostPortMap[HOST_PORT_CUR] {
+			// this will happen after controller restart
+			r.HostPortMap[HOST_PORT_CUR] = iport
+			return true
+		}
+	} else {
+		pdj.ObjectMeta.Annotations[hostPort] = fmt.Sprintf("%d", r.allocNewPort())
+		// bug fix
+		// c.f. https://github.com/elastic/cloud-on-k8s/issues/1822
+		if pdj.Spec.PS.Template.Spec.Containers == nil {
+			pdj.Spec.PS.Template.Spec.Containers = []corev1.Container{}
+		}
+		r.Update(ctx, pdj)
+		return true
+	}
+	return false
+}
+
+// allocNewPort globally
+func (r *PaddleJobReconciler) allocNewPort() int {
+	// all nodes should use the same port according to paddle launch
+	// so, alloc port by job instead by node now
+	return r.allocNewHostPort(HOST_PORT_CUR)
+}
+
+// allocNewHostPort alloc new hostport by key, key may refer to host, job, or global key
+func (r *PaddleJobReconciler) allocNewHostPort(key string) int {
+	if port, ok := r.HostPortMap[key]; ok {
+		if port+HOST_PORT_NUM < r.HostPortMap[HOST_PORT_END] {
+			r.HostPortMap[key] = port + HOST_PORT_NUM
+			return r.HostPortMap[key]
+		}
+	}
+	r.HostPortMap[key] = r.HostPortMap[HOST_PORT_START]
+	return r.HostPortMap[key]
+}
```

Comment:
Maybe just merge this function with `allocNewPort` into a single function? I feel the function doesn't make sense when `key` is not `HOST_PORT_CUR`.